src/sync.rs

```diff
@@ -0,0 +1,59 @@
+//! A module containing functions and utilities useful for synchronizing state.
+
+#![cfg_attr(not(all(target_vendor = "nintendo", target_env = "agb")), allow(unused_variables))]
+
+use crate::io::irq::{IrqEnableSetting, IME};
+
+mod locks;
+mod statics;
+
+pub use locks::*;
+pub use statics::*;
+
+/// Marks that a given pointer is read by volatile means without actually
+/// reading it.
+#[inline(always)]
+pub fn volatile_mark_ro<T>(val: *const T) {
+  unsafe { asm!("/* {0} */", in(reg) val, options(readonly, nostack)) }
+}
+
+/// Marks that a given pointer is read or written by volatile means without
+/// actually reading or writing it.
+#[inline(always)]
+pub fn volatile_mark_rw<T>(val: *mut T) {
+  unsafe { asm!("/* {0} */", in(reg) val, options(nostack)) }
+}
+
+/// An internal function used as a temporary hack to get `compiler_fence`
+/// working. While this call is not properly inlined, working is better than
+/// not working at all.
+///
+/// Ideally we should figure out why this causes any code to be emitted at all
+/// (probably some problem with our target JSON?) and fix it.
```

Comment:
Probably https://github.com/rust-lang/rust/issues/62256.

Reply:
Yeah. I found a stray comment in LLVM that looked similar, but I have no idea if this is the cause, or something else. Definitely something on the LLVM level, unfortunately. :c 

(https://github.com/llvm/llvm-project/blob/eb75f250feb6822d57be95e8535e28724cde6e9d/llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp#L3998)